### PR TITLE
allow cache and force refresh on table list

### PR DESCRIPTION
**Pull request by youngyjd**
This is a follow-up PR for https://github.com/apache/incubator-superset/pull/5933

1. Added `table_cache_timeout` settings to `metadata_cache_timeout` in database configuration. If `table_cache_timeout` is unset in the field, cache will not be enabled for the metadata fetch of this database. If it is set to 0, cache will never time out.
2. Added a button to sqllab UI which allows user to refetch table list metadata, or force refresh cache if cache is enabled.

screen shot 2018-10-11 at 4 53 56 pm

### CODE REVIEW

File: superset/assets/src/SqlLab/components/SqlEditorLeftBar.jsx [hunk 1]
```diff
@@ -41,12 +41,10 @@ class SqlEditorLeftBar extends React.PureComponent {
   onDatabaseChange(db, force) {
     const val = db ? db.value : null;
     this.setState({ schemaOptions: [] });
+    this.setState({ tableOptions: [] });
     this.props.actions.queryEditorSetSchema(this.props.queryEditor, null);
     this.props.actions.queryEditorSetDb(this.props.queryEditor, val);
-    if (!(db)) {
-      this.setState({ tableOptions: [] });
-    } else {
-      this.fetchTables(val, this.props.queryEditor.schema);
```
Comment (youngyjd) [inline superset/assets/src/SqlLab/components/SqlEditorLeftBar.jsx]: fetchTables here makes no sense because the database has been changed but `this.props.queryEditor.schema` is still the schema selected for the previously selected db. 
And it is set null on line 44:
`this.props.actions.queryEditorSetSchema(this.props.queryEditor, null);`

**Comment by youngyjd** [comment]
Please take a look @betodealmeida

File: superset/assets/src/SqlLab/components/SqlEditorLeftBar.jsx [hunk 2]
```diff
@@ -41,12 +41,10 @@ class SqlEditorLeftBar extends React.PureComponent {
   onDatabaseChange(db, force) {
     const val = db ? db.value : null;
     this.setState({ schemaOptions: [] });
+    this.setState({ tableOptions: [] });
```
Comment (mistercrunch) [inline superset/assets/src/SqlLab/components/SqlEditorLeftBar.jsx]: A single `setState` call with `({ schemaOptions: [], tableOptions: [] })` would be better as it would trigger a single render call instead of 2.

File: superset/assets/src/SqlLab/components/SqlEditorLeftBar.jsx [hunk 3]
```diff
@@ -196,7 +195,7 @@ class SqlEditorLeftBar extends React.PureComponent {
               <RefreshLabel
                 onClick={this.onDatabaseChange.bind(
                     this, { value: database.id }, true)}
-                tooltipContent="force refresh schema list"
+                tooltipContent="force refresh table list"
```
Comment (mistercrunch) [inline superset/assets/src/SqlLab/components/SqlEditorLeftBar.jsx]: shouldn't this say `schema`?

File: superset/assets/src/SqlLab/components/SqlEditorLeftBar.jsx [hunk 4]
```diff
@@ -214,30 +213,41 @@ class SqlEditorLeftBar extends React.PureComponent {
...
+                  onChange={this.changeTable.bind(this)}
+                  loadOptions={this.getTableNamesBySubStr.bind(this)}
+                />
+              }
+            </div>
+            <div className="col-md-1 col-xs-1" style={{ paddingTop: '8px', paddingLeft: '0px' }}>
+              <RefreshLabel
+                onClick={this.changeSchema.bind(
+                    this, { value: this.props.queryEditor.schema }, true)}
+                tooltipContent="force refresh schema list"
```
Comment (mistercrunch) [inline superset/assets/src/SqlLab/components/SqlEditorLeftBar.jsx]: shouldn't this say `table list`?

File: superset/assets/src/SqlLab/components/SqlEditorLeftBar.jsx (hunk 3 above)
Comment (mistercrunch) [reply superset/assets/src/SqlLab/components/SqlEditorLeftBar.jsx]: Also, don't forget `i18n` as in `tooltipContent={t('force refresh table list')}`

File: superset/assets/src/SqlLab/components/SqlEditorLeftBar.jsx (hunk 4 above)
Comment (mistercrunch) [reply superset/assets/src/SqlLab/components/SqlEditorLeftBar.jsx]: `i18n`

File: superset/db_engine_specs.py [hunk 5]
```diff
@@ -318,9 +318,25 @@ def get_schema_names(cls, inspector, db_id,
         return inspector.get_schema_names()
 
     @classmethod
-    def get_table_names(cls, schema, inspector):
+    @cache_util.memoized_func(
+        enable_cache=lambda *args, **kwargs: kwargs.get('enable_cache', False),
```
Comment (mistercrunch) [inline superset/db_engine_specs.py]: The lambda stuff looks a bit nasty to me, had to go read the decorator docstring to figure out what is going on here. I understand this is outside the scope of this PR so it's ok.

**Review by mistercrunch — CHANGES_REQUESTED** [review]
Few minor changes requested, otherwise LGTM

File: superset/assets/src/SqlLab/components/SqlEditorLeftBar.jsx (hunk 3 above)
Comment (youngyjd) [reply superset/assets/src/SqlLab/components/SqlEditorLeftBar.jsx]: woops. my bad.

**Review by mistercrunch — APPROVED** [review]
